File: Triangle.java
/*120. Triangle
Description  Submission  Solutions  Add to List
Total Accepted: 94001
Total Submissions: 287650
Difficulty: Medium
Contributors: Admin
Given a triangle, find the minimum path sum from top to bottom. Each step you may move to adjacent numbers on the row below.

For example, given the following triangle
[
     [2],
    [3,4],
   [6,5,7],
  [4,1,8,3]
]
The minimum path sum from top to bottom is 11 (i.e., 2 + 3 + 5 + 1 = 11).

Note:
Bonus point if you are able to do this using only O(n) extra space, where n is the total number of rows in the triangle.

Hide Tags Array Dynamic Programming
*/




/////////////////////////////////////////////////////////////////////////////////////
//C++




/*Bottom-up DP
space O(n)
in k th row:
path[k][i] = min(path[k+1][i], path[k+1][i+1]) + triangle[k][i]
simply set the path as a 1D array, and each time iteratively update itself
path[i] = min(path[i], path[k+1]) + triangle[k][i]
 */

/////////////////////////////////////////////////////////////////////////////////////
//Java
public class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int len = triangle.get(triangle.size()-1).size();
        int[] path = new int[len+1];
        for(int i = triangle.size(); i > 0; --i) //row, for the last row, extend the row + 1, only need i . 0
            for(int j = 0; j < i; ++j){ //col, only need j < i is enough,
                path[j] = Math.min(path[j], path[j+1])+triangle.get(i-1).get(j);
            }
        }
        return path[0];
    }
}

File: Triangle.swift
/*
Given a triangle array, return the minimum path sum from top to bottom.

For each step, you may move to an adjacent number of the row below. More formally, if you are on index i on the current row, you may move to either index i or index i + 1 on the next row.



Example 1:

Input: triangle = [[2],[3,4],[6,5,7],[4,1,8,3]]
Output: 11
Explanation: The triangle looks like:
   2
  3 4
 6 5 7
4 1 8 3
The minimum path sum from top to bottom is 2 + 3 + 5 + 1 = 11 (underlined above).
Example 2:

Input: triangle = [[-10]]
Output: -10


Constraints:

1 <= triangle.length <= 200
triangle[0].length == 1
triangle[i].length == triangle[i - 1].length + 1
-104 <= triangle[i][j] <= 104


Follow up: Could you do this using only O(n) extra space, where n is the total number of rows in the triangle?
*/

/*
Solution 3:
DP - Bottom Up

Time Complexity: O(n^2)
Space Complexity: O(n)
*/
class Solution3 {
    func minimumTotal(_ triangle: [[Int]]) -> Int {
        let n = triangle.count
        if n == 1 { return triangle[0][0] }
        var dp = triangle[n-1]
        for i in stride(from: n-2, through: 0, by: -1) {
            var temp = Array(repeating: 10001, count: n)
            for j in 0...i {
                let minJ = min(dp[j], dp[j+1]) + triangle[i][j]
                temp[j] = min(temp[j], minJ)
            }
            dp = temp
        }
        return dp[0]
    }
}

/*
Solution 2:
DP - Top Down

dp[i] means minimum sum final fall at col i
update dp n times for each row

Time Complexity: O(n^2)
Space Complexity: O(n)
*/
class Solution2 {
    func minimumTotal(_ triangle: [[Int]]) -> Int {
        let n = triangle.count
        if n == 1 { return triangle[0][0] }

        // dp[i]
        // minimum sum final fall at col i
        var dp = Array(repeating: 10001, count: n)
        dp[0] = triangle[0][0]

        for i in 1..<n {
            var temp = Array(repeating: 10001, count: n)
            for j in 0...i {
                temp[j] = min(temp[j], dp[j] + triangle[i][j])
                if j > 0 {
                    temp[j] = min(temp[j], dp[j-1] + triangle[i][j])
                }
            }
            dp = temp
        }

        var minSum = dp[0]
        for i in 1..<n {
            minSum = min(minSum, dp[i])
        }
        return minSum
    }
}

/*
Solution 1:
Time Limit Exceeded

Brute force
dfs

Time Complexity: O(2^n)
Space Complexity: O(1)
*/
class Solution1 {
    func minimumTotal(_ triangle: [[Int]]) -> Int {
        let n = triangle.count
        if n == 1 { return triangle[0][0] }

        var minSum = Int.max
        dfs(0, 0, n, triangle, triangle[0][0], &minSum)
        return minSum
    }

    func dfs(_ i: Int, _ j: Int, _ n: Int, _ triangle: [[Int]],
             _ curSum: Int, _ minSum: inout Int) {
        if i == n-1 {
            minSum = min(minSum, curSum)
            return
        }

        if i < n-1 {
            dfs(i+1, j, n, triangle, curSum+triangle[i+1][j], &minSum)
            if j <= i {
                dfs(i+1, j+1, n, triangle, curSum+triangle[i+1][j+1], &minSum)
            }
        }
    }
}